src/java.desktop/share/classes/javax/swing/plaf/basic/BasicDirectoryModel.java

```diff
@@ -412,6 +413,18 @@ private void cancelRunnables() {
                 runnable.cancel();
             }
         }
+
+        private synchronized <T> boolean compareIterators(Iterator<T> iterator1, Iterator<T> iterator2) {
```

Comment:
I'm not sure I understand, how this `synchronized` helps to avoid the issue.

Reply:
Since concurrent modification exception is thrown, it is clear that the List is being modified while comparing two list. Hence instead of copying the list locally, I have used iterators and comparing element by element in a `synchronized` method which ensures single thread is accessing the iterators. Without `synchronized` I guess it would again cause `concurrentModificationException`.

Reply:
Vector.iterator and Vector.subList.iterator are still check for modification on iteration (see usages of the method `java.util.AbstractList.SubList#checkForComodification`). It means, if vector was concurrently modified during iteration - iteration will fail with the `ConcurrentModificationException`

Reply:
Yes, which is why I am using Synchronized to handle `concurrentModificationException`.

Reply:
It doesn't work like this. Modification happen in another thread in another method. This `synchronized` doesn't affect another method.

Reply:
"_Synchronized methods enable a simple strategy for preventing thread interference and memory consistency errors: if an object is visible to more than one thread, all reads or writes to that object's variables are done through synchronized methods_" as per this statement from the doc ([Synchronized Methods](https://docs.oracle.com/javase/tutorial/essential/concurrency/syncmeth.html)) it should be handled right?

Reply:
_all reads or writes to that object's variables are done through synchronized methods_

This is a key requirement. In your case you have added `synchronized` only to method, which reads that object. But writer is not `synchronized` in you case.

Reply:
As per my analysis `fileCache` is synchronized in other places also. Do you notice any places where it is not synchronized?

Reply:
`synchronized` _works_ only if synchronized object is the same in all places. In the new method synchronized object - is an instance of `FilesLoader`, while in other places `fileCache` is accessed under `synchronized(fileCache)`.

Reply:
Then adding `synchronized(fileCache)` in `FilesLoader` where list comparison happens would be better idea?

Reply:
Yes. If you go with `synchronized` then it should be `synchronized (fileCache)`

Reply:
This is my concern as well. It looks as if synchronisation is missing where it's required and therefore it allows for concurrent modification.

Moreover, when you make a copy of `Vector`, you have to do it in a synchronised block with the `Vector` as the monitor: copying is iterating all the values, and you want to ensure it's done in an atomic way.

Therefore, as @turbanoff noted, equals must be invoked while you hold the object monitor.

It looks as if the previous fix didn't actually fix the problem, because `equals` is still called several times on `fileCache` or with it without proper synchronisation.

It seems the entire `call` that's passed to `ShellFolder.invoke` needs to be inside `synchronized (fileCache)` — its size, its elements are accessed many times while the method is executed. If another thread can modify `fileCache` concurrently, the assumptions inside the code could be broken. The method starts with `oldSize = fileCache.size()` and then iterates over elements in `fileCache` with `i < oldSize`. What if the number of elements in `fileCache` reduces and becomes less than `oldSize`.

Probably, the method should make its own local copy of `fileCache`, do its work using the unsynchronised local copy. When complete, perform another sanity check _inside a synchronized block_ at least to ensure the size hasn't changed and then 

The variables `newFileCache` and `newFiles` could be `ArrayList` instead of `Vector` to avoid synchronisation penalty where it's not needed.

Reply:
Using `synchronized (fileCache)` inside `ShellFolder.invoke` would be better and one solution right? Than making local copy and again doing sanity checks/changing to `ArrayList`?

Reply:
> Using `synchronized (fileCache)` inside `ShellFolder.invoke` would be better and one solution right? Than making local copy and again doing sanity checks/changing to `ArrayList`?

I'm afraid I can't answer this question without more details on what is achieved by this code. We need to look closely into what was done for [JDK-8240690](https://bugs.openjdk.org/browse/JDK-8240690): _Race condition between EDT and BasicDirectoryModel.FilesLoader.run0()_.

Putting the entire method into `synchronized (fileCache)` is an easy solution. Yet any other thread which accesses `fileCache` will be blocked until the code exits the synchronized block in the `call` method. This somewhat defeats updating the file list in the background, doesn't it?

Even if you'll go this route, I'm for replacing `Vector` with `ArrayList` for the `newFileCache` and `newFiles` variables. These are local variables, they're not accessed concurrently. Yet they're accessed from two threads: the current one and the one where `ShellFolder.invoke` runs, so there could be a need to use another synchronisation technique to ensure thread-safety between these two threads.

Reply:
> Even if you'll go this route, I'm for replacing `Vector` with `ArrayList` for the `newFileCache` and `newFiles` variables. These are local variables, they're not accessed concurrently. Yet they're accessed from two threads: the current one and the one where `ShellFolder.invoke` runs, so there could be a need to use another synchronisation technique to ensure thread-safety between these two threads.

Another Synchronization technique even after `synchronized (fileCache)`?

Reply:
> > Even if you'll go this route, I'm for replacing `Vector` with `ArrayList` for the `newFileCache` and `newFiles` variables. These are local variables, they're not accessed concurrently. Yet they're accessed from two threads: the current one and the one where `ShellFolder.invoke` runs, so there could be a need to use another synchronisation technique to ensure thread-safety between these two threads.
> 
> Another Synchronization technique even after `synchronized (fileCache)`?

Yes, even after.

Reply:
> > > Even if you'll go this route, I'm for replacing `Vector` with `ArrayList` for the `newFileCache` and `newFiles` variables. These are local variables, they're not accessed concurrently. Yet they're accessed from two threads: the current one and the one where `ShellFolder.invoke` runs, so there could be a need to use another synchronisation technique to ensure thread-safety between these two threads.
> > 
> > 
> > Another Synchronization technique even after `synchronized (fileCache)`?
> 
> Yes, even after.

No, it's not needed, actually. `ShellFolder.invoke` executes the code directly on the same thread or passes it to COM thread on Windows. When it's passed to the COM thread, a `Future` object is used, which ensures synchronisation.

I submitted [JDK-8324973](https://bugs.openjdk.org/browse/JDK-8324973): _Replace Vector with ArrayList in BasicDirectoryModel.FilesLoader_.